src/strings_internal.h

```diff
@@ -0,0 +1,123 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file strings_interal.h Types and functions related to the internal workings of formatting OpenTTD's strings. */
+
+#ifndef STRINGS_INTERNAL_H
+#define STRINGS_INTERNAL_H
+
+/**
+ * Equivalent to the std::back_insert_iterator in function, with some
+ * convenience helpers for string concatenation.
+ *
+ * The formatter is currently backed by an external char buffer, and has some
+ * extra functions to ease the migration from char buffers to std::string.
+ */
+class StringBuilder {
+	char *current; ///< The current location to add strings
+	const char *last; ///< The last element of the buffer.
+
+public:
+	/* Required type for this to be an output_iterator; mimicks std::back_insert_iterator. */
+	using value_type = void;
+	using difference_type = void;
+	using iterator_category = std::output_iterator_tag;
+	using pointer = void;
+	using reference = void;
+
+	/**
+	 * Create the builder of an external buffer.
+	 * @param start The start location to write to.
+	 * @param last  The last location to write to.
+	 */
+	StringBuilder(char *start, const char *last) : current(start), last(last) {}
+
+	/* Required operators for this to be an output_iterator; mimicks std::back_insert_iterator. */
+	StringBuilder &operator++() { return *this; }
+	StringBuilder operator++(int) { return *this; }
+	StringBuilder &operator*() { return *this; }
+
+	/**
+	 * Operator to add a character to the end of the buffer. Like the back
+	 * insert iterators this also increases the position of the end of the
+	 * buffer.
+	 * @param value The character to add.
+	 * @return Reference to this inserter.
+	 */
+	StringBuilder &operator=(const char value)
+	{
+		return this->operator+=(value);
+	}
+
+	/**
+	 * Operator to add a character to the end of the buffer.
+	 * @param value The character to add.
+	 * @return Reference to this inserter.
+	 */
+	StringBuilder &operator+=(const char value)
+	{
+		if (this->current != this->last) *this->current++ = value;
+		return *this;
+	}
+
+	/**
+	 * Operator to append the given string to the output buffer.
+	 * @param str The string to add.
+	 * @return Reference to this inserter.
+	 */
+	StringBuilder &operator+=(const char *str)
+	{
+		this->current = strecpy(this->current, str, this->last);
+		return *this;
+	}
+
+	/**
+	 * Operator to append the given string to the output buffer.
+	 * @param str The string to add.
+	 * @return Reference to this inserter.
+	 */
+	StringBuilder &operator+=(const std::string &str)
+	{
+		return this->operator+=(str.c_str());
+	}
+
+	/**
+	 * Encode the given Utf8 character into the output buffer.
+	 * @param c The character to encode.
+	 * @return true iff there was enough space and the character got added.
+	 */
+	bool Utf8Encode(WChar c)
```

Comment:
Why does this as only function return a book? And of the three cases I saw above, only 2 check thee result and the third doesn't? Shouldn't this just silently ignore adding like the rest does?

See (1)

Reply:
It returns a bool so it can remain functionally the same as it was before. In the future, when we move to std::string, no function will (silently) truncate the string anymore, so this can return void and even `Remaining()` can go away. But not in this step; first move without functional changes to something where we can replace the `char[]` with `std::string`.

The case that doesn't check is due to me performing some "optimisations"; otherwise you would have:
```
void function(some params) {
   ...
   if (negative) {
      if (!builder.Utf8Encode(SCC_POP_COLOUR)) return;
   }
   /* Implicit return due to function ending. */
}
```

Reply:
What I find a bit odd about this choice, is that you only did it for `Utf8Encode`. Things like `+=` etc did change functional, and do silently drop the addition. So what is it that makes `Utf8Encode` special, is basically what I was asking. Don't any of the other cases need to return early (for now) if the string was full?

Reply:
To put Discord here, I took another look at the code: the rest already silently ignored when a string was full. So that is unchanged with your commits.

I guess they did check the length and returned early for `Utf8Encode`, as otherwise you can end up in situations where you inserted only part of the codepoint in the string, which is bad. But you don't do that anyway. And when I look at the code, if the string is full after `SCC_RED`, it will happily run the rest of the function, trying to keep on adding to the string.

So I would suggest to already extend that to this function. If `SCC_PUSHCOLOR` already doesn't fit: who cares, let it run on for those 2 more functions.

Already prepares you for the future, and does away with the special exception :) My 2 cents .. I can't see the whole picture here ofc :)

Reply:
And now I talked myself in a corner ... `SCC_RED` might not fit, but the `-` added after that might. So that would be very wrong. No I take it back, your code is correct.

Reply:
No, `+=` does not change functionally as it replaces a `strecpy` with a (now hidden) `strecpy`. Both truncate(d) silently.

`Utf8Encode` is a different beast, as that function in itself is not even doing bounds checking. So, in many places where it is used there is a "is there enough space to put this encoded character" check, which does then decide what to do. I moved that "is there enough space" check into `Utf8Encode` and exposed the result, so things can remain functionally the same while performing the migration.